drivers/crypto/crypto_it8xxx2_sha_v2.c

```diff
@@ -0,0 +1,280 @@
+
+static void it8xxx2_sha256_init(bool init_k)
+{
+	chip_ctx.sha_init = init_k;
+	chip_ctx.w_input_index = 0;
+	chip_ctx.total_len = 0;
+
+	/* Set DLM address for input data */
+	sys_write8(((uint32_t)&chip_ctx) & 0xc0,
+			IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHADBA0R);
+	sys_write8(((uint32_t)&chip_ctx) >> 8,
+			IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHADBA1R);
+}
+
+static void it8xxx2_sha256_module_calculation(void)
+{
+	struct gctrl_it8xxx2_regs *const gctrl_regs = GCTRL_IT8XXX2_REGS_BASE;
+	uint32_t key;
+	uint8_t sha_ctrl;
+
+	sha_ctrl = sys_read8(IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHACR);
+
+	if (chip_ctx.sha_init) {
+		sha_ctrl |= (IT8XXX2_SHAINI | IT8XXX2_SHAEXE);
+		chip_ctx.sha_init = 0;
+	} else {
+		sha_ctrl |= IT8XXX2_SHAEXE;
+	}
+
+	/*
+	 * Global interrupt is disabled because the CPU cannot access memory
+	 * via the DLM (Data Local Memory) bus while HW module is computing
+	 * hash.
+	 */
+	key = irq_lock();
+	/* Crypto use SRAM */
+	gctrl_regs->GCTRL_PMER3 |= IT8XXX2_GCTRL_SRAM_CRYPTO_USED;
+	sys_write8(sha_ctrl, IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHACR);
+
+	while (sys_read8(IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHASR) & IT8XXX2_SHABUSY)
+		;
+
+	sys_write8(IT8XXX2_SHAIS, IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHASR);
+	/* CPU use SRAM */
+	gctrl_regs->GCTRL_PMER3 &= ~IT8XXX2_GCTRL_SRAM_CRYPTO_USED;
+	gctrl_regs->GCTRL_PMER3;
+	irq_unlock(key);
+
+	chip_ctx.w_input_index = 0;
+}
+
+static int it8xxx2_hash_handler(struct hash_ctx *ctx, struct hash_pkt *pkt,
+				bool finish)
+{
+	struct gctrl_it8xxx2_regs *const gctrl_regs = GCTRL_IT8XXX2_REGS_BASE;
+	uint32_t rem_len = pkt->in_len;
+	uint32_t in_buf_idx = 0;
+	uint32_t i, key;
+	uint8_t sha_ctrl;
+
+	while (rem_len) {
+		/* Data length >= 1KB */
+		if (rem_len >= SHA_SHA256_SRAM_BUF) {
+			rem_len = rem_len - SHA_SHA256_SRAM_BUF;
+
+			for (i = 0; i < SHA_SHA256_SRAM_BUF; i++) {
+				chip_ctx.w_input[chip_ctx.w_input_index++] =
+								pkt->in_buf[in_buf_idx++];
+			}
+
+			/* HW automatically load 1KB data from DLM */
+			sys_write8(IT8XXX2_SHAEXEC_1KByte,
+					IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHAECR);
+			it8xxx2_sha256_module_calculation();
+		} else {
+			/* 0 <= Data length < 1KB */
+			while (rem_len) {
+				rem_len--;
+				chip_ctx.w_input[chip_ctx.w_input_index++] =
+								pkt->in_buf[in_buf_idx++];
+
+				if (chip_ctx.w_input_index >= SHA_SHA256_BLOCK_LEN) {
+					/* HW automatically load 64Bytes data from DLM */
+					sys_write8(IT8XXX2_SHAEXEC_64Byte,
+							IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHAECR);
+					it8xxx2_sha256_module_calculation();
+				}
+			}
+		}
+	}
+
+	chip_ctx.total_len += pkt->in_len;
+
+	if (finish) {
+		uint32_t *ob_ptr = (uint32_t *)pkt->out_buf;
+
+		/* Pre-processing (Padding) */
+		memset(&chip_ctx.w_input[chip_ctx.w_input_index],
+			0, SHA_SHA256_BLOCK_LEN - chip_ctx.w_input_index);
+		chip_ctx.w_input[chip_ctx.w_input_index] = 0x80;
+
+		if (chip_ctx.w_input_index >= 56) {
```

Comment:
Why 56 bytes here and not 64?

Reply:
Because the last two bytes are bit length field. If the data index >= 56,  it needs to trigger HW to calculate, then fill 0 data and the last two bytes bit length and calculate again.
This `if (chip_ctx.w_input_index >= 56) {} `handles the boundary case of rest data.

Reply:
Please add a comment explaining this code path.